autobahn/autobahn/wamp/protocol.py

```diff
@@ -481,15 +478,20 @@ def error(err):
                     raise ProtocolError("PUBLISHED received for non-pending request ID {0}".format(msg.request))
 
             elif isinstance(msg, message.Subscribed):
-
                 if msg.request in self._subscribe_reqs:
                     d, obj, fn, topic, options = self._subscribe_reqs.pop(msg.request)
-                    if options:
-                        self._subscriptions[msg.subscription] = Handler(obj, fn, topic, options.details_arg)
-                    else:
-                        self._subscriptions[msg.subscription] = Handler(obj, fn, topic)
-                    s = Subscription(self, msg.subscription)
-                    self._resolve_future(d, s)
+                    sub_id = msg.subscription
+                    # we should NEVER have our ID subscribed already
```

Comment:
It should be a list of handlers. It should do the SUBSCRIBE, and when the subscription ID comes back look into the list and append the handler if it's not already in the list. When doing unsubscribe(handler) it should remove the given handler (if it's in the list) and if the list becomes empy, send out an UNSUBSCRIBE. If doing an unqualified unsubscribe() - that is one not providing a specific handler to be removed - it should just send out the UNSUBSCRIBE (and remove the list from the map).

The subscription ID will be the same for the pair (topic, match policy). Two subscribes for the same topic, but different match policies will return different subscription IDs.